csv.go

```diff
@@ -0,0 +1,180 @@
+	longestUptimeEnd := tcping.longestUptime.end.Format(csvTimeFormat)
+	longestDowntimeDuration := tcping.longestDowntime.duration.String()
+	longestDowntimeStart := tcping.longestDowntime.start.Format(csvTimeFormat)
+	longestDowntimeEnd := tcping.longestDowntime.end.Format(csvTimeFormat)
+
+	totalDuration := time.Since(tcping.startTime).String()
+	if !tcping.endTime.IsZero() {
+		totalDuration = tcping.endTime.Sub(tcping.startTime).String()
+	}
+
+	row := []string{
+		"statistics",
+		time.Now().Format(csvTimeFormat),
+		tcping.userInput.ip.String(),
+		tcping.userInput.hostname,
+		strconv.Itoa(int(tcping.userInput.port)),
+		strconv.Itoa(int(tcping.retriedHostnameLookups)),
+		strconv.Itoa(int(tcping.totalSuccessfulProbes)),
+		strconv.Itoa(int(tcping.totalUnsuccessfulProbes)),
+		strconv.FormatBool(tcping.lastSuccessfulProbe.IsZero()),
+		strconv.FormatBool(tcping.lastUnsuccessfulProbe.IsZero()),
+		lastSuccessfulProbe,
+		lastUnsuccessfulProbe,
+		strconv.Itoa(int(totalPackets)),
+		fmt.Sprintf("%.2f", packetLoss),
+		tcping.totalUptime.String(),
+		tcping.totalDowntime.String(),
+		longestUptimeDuration,
+		longestUptimeStart,
+		longestUptimeEnd,
+		longestDowntimeDuration,
+		longestDowntimeStart,
+		longestDowntimeEnd,
+		fmt.Sprintf("%.3f", tcping.rttResults.min),
+		fmt.Sprintf("%.3f", tcping.rttResults.average),
+		fmt.Sprintf("%.3f", tcping.rttResults.max),
+		tcping.startTime.Format(csvTimeFormat),
+		tcping.endTime.Format(csvTimeFormat),
+		totalDuration,
+	}
+
+	return cp.writer.Write(row)
+}
+
+func (cp *csvPrinter) saveHostNameChange(h []hostnameChange) error {
+	for _, host := range h {
+		if !host.Addr.IsValid() {
+			continue
+		}
+		row := []string{
+			"hostname change",
+			host.When.Format(csvTimeFormat),
+			host.Addr.String(),
+			"", // Empty fields for consistency with stats rows
+			"", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "", "",
+		}
+		spew.Dump(row)
+		if err := cp.writer.Write(row); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (cp *csvPrinter) printStart(hostname string, port uint16) {
+	fmt.Printf("TCPinging %s on port %d\n", hostname, port)
+}
+
+func (cp *csvPrinter) printStatistics(tcping tcping) {
+	err := cp.saveStats(tcping)
+	if err != nil {
+		cp.printError("\nError while writing stats to the CSV file %q\nerr: %s", cp.filename, err)
+	}
+
+	if !tcping.endTime.IsZero() {
+		err = cp.saveHostNameChange(tcping.hostnameChanges)
+		if err != nil {
+			cp.printError("\nError while writing hostname changes to the CSV file %q\nerr: %s", cp.filename, err)
+		}
+	}
+
+	cp.writer.Flush()
+	if err := cp.writer.Error(); err != nil {
+		cp.printError("Error flushing CSV writer: %v", err)
+	}
+
+	fmt.Printf("\nStatistics for %q have been saved to %q\n", tcping.userInput.hostname, cp.filename)
+}
+
+func (cp *csvPrinter) printError(format string, args ...interface{}) {
+	fmt.Fprintf(os.Stderr, format+"\n", args...)
```

Comment:
why not use the already existing printError here?

https://github.com/pouriyajamshidi/tcping/blob/24ebf6a286e8988acde0df262d31a7c3dae69803/statsprinter.go#L219